### PR TITLE
HUM-180 Support for proxy handoffs on writes

### PR DESCRIPTION
**Pull request by redbo**

### CODE REVIEW

File: client/directclient.go [hunk 1]
```diff
@@ -80,58 +82,170 @@ func NewProxyDirectClient(policyList conf.PolicyList) (*ProxyDirectClient, error
...
-	for responseCount < len(reqs) {
+	for q.outstanding() {
 		select {
-		case <-responses:
-			responseCount++
+		case response := <-responses:
+			q.addResponse(response.i, response.r)
 		case <-timeout:
-			break waiting
+			return
```
Comment (corystone) [inline client/directclient.go]: This should still go back to the loop right?

Comment (redbo) [inline client/directclient.go]: Go back to the loop?  No, it was breaking out of the loop to hit the return below.  But just returning here let me get rid of the loop label.

File: client/directclient.go [hunk 2]
```diff
@@ -83,67 +85,205 @@ func NewProxyDirectClient(policyList conf.PolicyList) (*ProxyDirectClient, error
...
+}
+
+func (q *stdQuorumer) quorumMet(writers int) bool {
+	if writers >= q.replicaCount {
+		return true
+	}
+	if writers >= q.q && (len(q.responses)+writers) >= q.replicaCount*2 {
+		return true
+	}
+	for i := 2; i < 5; i++ {
```
Comment (gholt) [inline client/directclient.go]: Same here as above, though obviously I know better what's going on now.

File: client/directclient.go [hunk 3]
```diff
@@ -83,67 +85,205 @@ func NewProxyDirectClient(policyList conf.PolicyList) (*ProxyDirectClient, error
...
+}
+
+func (q *stdQuorumer) getResponse(timeout time.Duration) *http.Response {
+	end := time.After(timeout)
+	for {
+		// see if quorum has already been met
+		for _, r := range q.responses {
+			if q.rc[r.StatusCode/100] >= q.q {
+				// Give pending requests a chance to finish, to improve consistency of read-after-write.
+				end := time.After(PostQuorumTimeoutMs * time.Millisecond)
```
Comment (gholt) [inline client/directclient.go]: I dislike shadowed vars, but just a suggestion to rename this "end" to something else.

File: client/directclient.go [hunk 4]
```diff
@@ -83,67 +85,205 @@ func NewProxyDirectClient(policyList conf.PolicyList) (*ProxyDirectClient, error
...
+		// bail out if quorum isn't possible
+		quorumPossible := false
+		for _, c := range q.rc {
+			if (c+q.requestCount)-len(q.responses) >= q.q {
+				quorumPossible = true
+			}
+		}
+		if !quorumPossible {
+			return ResponseStub(http.StatusServiceUnavailable, "The service is currently unavailable.")
+		}
```
Comment (gholt) [inline client/directclient.go]: The above code is fine, but it does offer the immediate question of "why not use the quorumPossible() func?" Probably should rename the quorumPossible func to launchUntilQuorumPossible or something.

File: client/directclient.go [hunk 5]
```diff
@@ -83,67 +85,205 @@ func NewProxyDirectClient(policyList conf.PolicyList) (*ProxyDirectClient, error
...
+		var dev *ring.Device
+		if requestCount < len(devs) {
+			dev = devs[requestCount]
+		} else {
+			dev = more.Next()
+			if dev == nil {
+				break
+			}
+		}
+		fmt.Println(requestCount, dev)
```
Comment (gholt) [inline client/directclient.go]: Leftover debug?

File: client/directclient.go [hunk 6]
```diff
@@ -596,140 +742,179 @@ func newObjectClient(proxyDirectClient *ProxyDirectClient, account string, conta
 	return &standardObjectClient{proxyDirectClient: proxyDirectClient, account: account, container: container, policy: ci.StoragePolicyIndex, objectRing: objectRing}
 }
 
+// putReader is a Reader proxy that calls its notify() function the first time Read is called.
```
Comment (gholt) [inline client/directclient.go]: I don't think there is a notify() func.

Comment (redbo) [inline client/directclient.go]: Well there used to be!

File: client/directclient.go (hunk 5 above)
Comment (redbo) [inline client/directclient.go]: Yes

File: client/directclient.go (hunk 4 above)
Comment (redbo) [inline client/directclient.go]: Okay.

File: client/directclient.go (hunk 3 above)
Comment (redbo) [inline client/directclient.go]: Okay

File: client/directclient.go [hunk 7]
```diff
@@ -83,67 +85,205 @@ func NewProxyDirectClient(policyList conf.PolicyList) (*ProxyDirectClient, error
...
+		q.failures = append(q.failures, index)
+	} else if resp.StatusCode >= 0 {
+		q.rc[resp.StatusCode/100]++
+	}
+}
+
+func (q *stdQuorumer) quorumPossible() bool {
+	// if quorum has become impossible, launch requests until it's possible again.
+	for q.requestCount < q.replicaCount*2 {
+		for i := 2; i < 5; i++ {
```
Comment (redbo) [inline client/directclient.go]: Will do.  but responseClass and responseClassCount.

**Pull request merged**